src/sources/host_metrics.rs

```diff
@@ -0,0 +1,1100 @@
+                                    tags.clone(),
+                                ),
+                                self.counter(
+                                    "disk_reads_completed_total",
+                                    timestamp,
+                                    counter.read_count() as f64,
+                                    tags.clone(),
+                                ),
+                                self.counter(
+                                    "disk_written_bytes_total",
+                                    timestamp,
+                                    counter.write_bytes().get::<byte>() as f64,
+                                    tags.clone(),
+                                ),
+                                self.counter(
+                                    "disk_writes_completed_total",
+                                    timestamp,
+                                    counter.write_count() as f64,
+                                    tags,
+                                ),
+                            ]
+                            .into_iter(),
+                        )
+                    })
+                    .flatten()
+                    .collect::<Vec<_>>()
+                    .await
+            }
+            Err(error) => {
+                error!(message = "Failed to load disk I/O info", %error, rate_limit_secs = 60);
+                vec![]
+            }
+        }
+    }
+
+    fn counter(
+        &self,
+        name: &str,
+        timestamp: DateTime<Utc>,
+        value: f64,
+        tags: BTreeMap<String, String>,
+    ) -> Metric {
+        Metric {
+            name: self.namespace.encode(name),
+            timestamp: Some(timestamp),
+            kind: MetricKind::Absolute,
+            value: MetricValue::Counter { value },
+            tags: Some(tags),
+        }
+    }
+
+    fn gauge(
+        &self,
+        name: &str,
+        timestamp: DateTime<Utc>,
+        value: f64,
+        tags: BTreeMap<String, String>,
+    ) -> Metric {
+        Metric {
+            name: self.namespace.encode(name),
+            timestamp: Some(timestamp),
+            kind: MetricKind::Absolute,
+            value: MetricValue::Gauge { value },
+            tags: Some(tags),
+        }
+    }
+}
+
+async fn filter_result<T>(result: Result<T, Error>, message: &'static str) -> Option<T> {
+    result
+        .map_err(|error| error!(message, %error, rate_limit_secs = 60))
+        .ok()
+}
+
+fn vec_contains_path(vec: &Option<Vec<PatternWrapper>>, value: &Path) -> Option<()> {
+    match vec {
+        // No patterns list matches everything
+        None => Some(()),
+        // Otherwise find the given value
+        Some(vec) => vec
+            .iter()
+            .find(|&pattern| pattern.matches_path(value))
+            .map(|_| ()),
+    }
+}
+
+fn vec_contains_str(vec: &Option<Vec<PatternWrapper>>, value: &str) -> Option<()> {
```

Comment:
I see this essentially `or`s the filters together?

I'm wondering if we wouldn't expect negations to filter out values that previously matched though. For example, if I have a filter `[["sd*", "!sda"]` would the user expect "sda" to be rejected? I personally would have.

Or, if I had `["sd*", "!dm-*"]` I might have expected "xvda" to be rejected, but I think this would allow that through as it matches "!dm-*".

We may want to structure it as `or`ing all positive filters (i.e. not negations) and then applying the negations; however, that would require a filter set like `["*", "!dm-*"]` to literally grab anything that isn't `dm-*`.

We could also consider providing a simple filter syntax that would let people write things like `"sd*" && !sda && !dm-*" . I'm reminded of `tcpdump`'s filter syntax (e.g. `tcpdump -i eth1 port not 22 and port 53 and host 1.2.3.4`). That would require a bit more heavy lifting though :smile: . We could also just drop negations for now and only allow positive matches.

Reply:
This does "or" all the filters together. The first one to match passes the filter. I'm not excited about making the filter logic more complex without a demonstrated use case, as explaining how it works will be a pain.

One alternative would be to split the filter lists into pairs of lists. That is:
```toml
disk.devices.include = ["d*"]
disk.devices.exclude = ["dm-*"]
```

That would be easy to explain and straightforward to implement. I'm not particularly proud of the filter negation mechanism, but it was part of the RFC. @jamtur01 any opinion?

Reply:
The alternative seems reasonable. I guess it's easier to document?

Reply:
:+1: I like splitting it up into `include` and `exclude`. That seems relatively simple from  the user's perspective.